### PR TITLE
WIP: Start creating a builder for the StorageClient.

### PR DESCRIPTION
**Pull request by gorzell**
This is a follow on to the work done in #964 that implements a similar pattern for the `StorageClient`. One of the main benefits in addition to consistency is that it allows for configuring more of the client options like `retries`, etc.

### CODE REVIEW

File: sdk/storage/src/core/clients/storage_client.rs [hunk 1]
```diff
@@ -66,6 +66,114 @@ pub enum ServiceType {
...
+
+    #[must_use]
+    pub fn access_key(mut self, key: impl Into<String>) -> Self {
+        match self.storage_credentials {
+            None => {
+                self.storage_credentials =
+                    Some(StorageCredentials::Key(self.account.clone(), key.into()));
+                self
+            }
+            // FIXME: This gets a bit tricky in the emulator case...
```
Comment (rylev) [inline sdk/storage/src/core/clients/storage_client.rs:133]: The way we're handling this in Cosmos is by not having an `access_key` (or in the case of Cosmos an `AuthorizationToken`) setter. As you rightfully pointed out here, the need for authentication comes down to whether the user is talking to an emulator or not. 

This is why this is [all encapsulated in the `CloudLocation`](https://github.com/Azure/azure-sdk-for-rust/blob/bc1d085a8d6ba864a7acc0cccbad78efef348561/sdk/data_cosmos/src/clients/cosmos.rs#L179-L197). I think we'll want a similar construct here as well.

File: sdk/storage/src/core/clients/storage_client.rs [hunk 2]
```diff
@@ -66,6 +66,114 @@ pub enum ServiceType {
...
+            Some(c) => panic!(
+                "credentials already set to: {:?}, you should only set them once.",
+                c
+            ),
```
Comment (rylev) [inline sdk/storage/src/core/clients/storage_client.rs:137]: Is there a need for panicking here? Why not let the user override credentials? We allow them to override other items that might already be set.

File: sdk/storage/src/core/clients/storage_client.rs [hunk 3]
```diff
@@ -66,6 +66,114 @@ pub enum ServiceType {
...
+    storage_options: StorageOptions,
+}
+
+impl StorageClientBuilder {
+    #[must_use]
+    pub fn new(account: impl Into<String>) -> Self {
+        let account = account.into();
+        Self {
+            storage_credentials: None,
+            blob_storage_url: get_endpoint_uri(None, &account, "blob").unwrap(),
```
Comment (rylev) [inline sdk/storage/src/core/clients/storage_client.rs:87]: I think we'll likely want to rework how we handle urls completely. In Cosmos, we ask the [`CloudLocation` for the right url]((https://github.com/Azure/azure-sdk-for-rust/blob/bc1d085a8d6ba864a7acc0cccbad78efef348561/sdk/data_cosmos/src/clients/cosmos.rs#L132) (more on `CloudLocation` below). 

This should also work here, but we'll just need to include one additional piece of information: the resource type. We can encapsulate that in an enum which gets passed to the `url` method which can then construct the right url based on cloud location and the resource type. 

I think for now, we can just construct the url as a new `String` every time. Once we get an API we're happy with we can add back some sort of memoization should we need to.

**Review by rylev — CHANGES_REQUESTED** [review]
This is a good start, but I think unfortunately, we'll need a more fundamental refactoring here to get to where we want to go.

**Comment by cataggar** [comment]
@gorzell, we are preparing a release very soon #994. In case you wish to squeeze this in, let me know.

**Comment by rylev** [comment]
This has been superseded by #1057 and subsequent PRs.

**Pull request closed without merging**